### DISCUSSION TITLE
Conditionally running a git hook or command

### DISCUSSION THREAD

**Discussion by forrestab** [alirezanet/Husky.Net#45] [Q&A] [answered]
I want to be able to run my integration tests (I know not the best, but I have reasons) before code is committed, but only if any of the staged files meet a certain condition. I tried creating a csx file and passing the `${staged}` variable in, then check for the condition and Im able to conditionally run `dotnet test`. The problem that Im running into is how to fail the commit if a test fails. Here is what I have so far:

```
// task

{
   "name": "integration-tests-runner",
   "group": "pre-commit",
   "command": "dotnet",
   "args": [ "husky", "exec", ".husky/csx/run-integration-tests.csx", "--args", "${staged}" ]
}

// pre-commit hook

#!/bin/sh
. "$(dirname "$0")/_/husky.sh"

dotnet husky run -p --group "pre-commit" --args "$1"

// csx

#r "System.Linq"

using System.Diagnostics;
using System.Linq;

if (Args.Any(stagedFile => stagedFile.Contains("src/"))) 
{
    Process.Start("dotnet", "test src/IntegrationTests/");
}
```

If a test fails, the exception is shown in the terminal, but the commit goes through. 

This may be outside the scope of this package and thats fine, but I wanted to make sure there wasnt an easier way since this is the first time Ive worked with csx files.

**Comment by alirezanet** [comment, marked as answer]
Hi @forrestab,

In the CSX files, you can return an **exception** or **non-zero exit code** to break the execution (e.g `if (some condition) return 1`),

```c#
var x = Process.Start("dotnet", "test src/IntegrationTests/");
x.WaitForExit();
return x.ExitCode;
```

but this is an interesting problem, I'll look for an easier way to handle this soon, feel free to open an issue if you have any suggestions to make this operation easier.

**Reply by forrestab** [reply]
Not sure if its the intended purpose, but `task.include` gave me the conditional execution of the hook that I was looking for. I simply specified the file extensions that would need to tests to run (i.e. .csproj, .cs, etc). If those files arent staged, then the hook doesn't fire with a message, "No files found" or something similar.

As far as running the `dotnet test` command in the hook, a design decision was made with that command that it does not return anything but 0, regardless of whether or not tests fail. So I ended up having to part the test results file and if the outcome is "Failed" then I return 1 otherwise 0.

@alirezanet I really appreciate all of your help with my various questions.

**Reply by alirezanet** [reply]
Yes `include` and `exclude` both can be used for filtering files. I wanted to suggest you to try these attributes but I assumed you might have a more complex condition.

I'm glad that you have solved the problem. don't hesitate if you had any other questions, these discussions might also help others that face a similar problem.
good luck.
